indexes: Don't wipe indexes again when continuing a prior reindex

sedited opened the pull request:
When restarting `bitcoind` during an ongoing reindex without setting the `-reindex` flag again, the block and coins db is left intact, but any data from the optional indexes is discarded. While not a bug per se, wiping the data again is
wasteful, both in terms of having to write it again,  as well as potentially leading to longer startup times. So keep the  index data instead when continuing a prior reindex.

Also includes a bugfix and smaller code cleanups around the reindexing code. The bug was introduced in b47bd959207e82555f07e028cc2246943d32d4c3: "kernel: De-globalize fReindex".

sedited committed:
bugfix: Streamline setting reindex option

Reverts a bug introduced in b47bd959207e82555f07e028cc2246943d32d4c3
"kernel: De-globalize fReindex". The change leads to a GUI user being
prompted to re-index on a chainstate loading failure more than once as
well as the node actually not reindexing if the user chooses to. Fix
this by setting the reindexing option instead of the atomic, which can
be safely re-used to indicate that a reindex should be attempted.

The bug specifically is caused by the chainman, and thus the blockman
and its m_reindexing atomic being destroyed on every iteration of
the for loop.

The reindex option for ChainstateLoadOptions is currently also set in a
confusing way. By using the reindex atomic, it is not obvious in which
scenario it is true or false.

The atomic is controlled by both the user passing the -reindex option,
the user chosing to reindex if something went wrong during chainstate
loading when running the gui, and by reading the reindexing flag from
the block tree database in LoadBlockIndexDB. In practice this read is
done through the chainstate module's CompleteChainstateInitialization's
call to LoadBlockIndex. Since this is only done after the reindex option
is set already, it does not have an effect on it.

Make this clear by using the reindex option from the blockman opts which
is only controlled by the user.

sedited committed:
indexes: Don't wipe indexes again when already reindexing

Before this change continuing a reindex without the -reindex flag set
would leave the block and coins db intact, but discard the data of the
optional indexes. While not a bug per se, wiping the data again is
wasteful, both in terms of having to write it again, and potentially
leading to longer startup times.

When initially running a reindex, both the block index and any further
activated indexes are wiped. On an index's Init(), both the best block
stored by the index and the chain's tip are null. An index's m_synced
member is therefore true. This means that it will process blocks through
validation events while the reindex is running.

Currently, if the reindex is continued without the user re-specifying
the reindex flag, the block index is preserved but further index data is
wiped. This leads to the stored best block being null, but the chain tip
existing. The m_synced member will be set to false. The index will not
process blocks through the validation interface, but instead use the
background sync once the reindex is completed.

If the index is preserved (this change) after a restart its best block
may potentially match the chain tip. The m_synced member will be set to
true and the index can process validation events during the rest of the
reindex.

sedited committed:
validation: Remove needs_init from LoadBlockIndex

It does not control any actual logic and the log message as well as the
comment are obsolete, since no database initialization takes place there
anymore. Log messages indicating when indexes and chainstate databases
are loaded exist in other places.

stickies-v commented:
Concept ACK

stickies-v commented on src/init.cpp:
```diff
@@ -1639,17 +1639,17 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     if (args.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
-        g_txindex = std::make_unique<TxIndex>(interfaces::MakeChain(node), cache_sizes.tx_index, false, chainman.m_blockman.m_reindexing);
+        g_txindex = std::make_unique<TxIndex>(interfaces::MakeChain(node), cache_sizes.tx_index, false, blockman_opts.reindex);
```
As discussed offline, I think this change means we're ignoring the GUI-based reindex from  https://github.com/bitcoin/bitcoin/pull/30132/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1609, both for these optional indices but also for the block index

TheCharlatan commented:
Updated 991f50acad6809c83c0c44cf3c55fb9d2c0107e3 -> 9de8b263dabd6dd2f86f1f0253c6ee3fac7a4407 ([preserveIndexOnRestart_0](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_0) -> [preserveIndexOnRestart_1](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/preserveIndexOnRestart_0..preserveIndexOnRestart_1))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1606818901), fixing a bug introduced in b47bd959207e82555f07e028cc2246943d32d4c3: "kernel: De-globalize fReindex".

ryanofsky commented:
> fixing a bug introduced in b47bd959207e82555f07e028cc2246943d32d4c3: "kernel: De-globalize fReindex".

Is this true? That commit, which was part of #29817, should not have changed any previous behavior

furszy reviewed (commented):
Looking good in a first glance. It would be nice to add some coverage for it just so it doesn't happen again. Maybe assert that certain logs are not present during init? Like the "Wiping LevelDB in <index_path>" one.

stickies-v commented:
Approach ACK. First 2 commits (9de8b263dabd6dd2f86f1f0253c6ee3fac7a4407) LGTM but the third one I'm going to need to spend a lot more time wrapping my head around the implications.

sedited committed:
test: Add functional test for continuing a reindex

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

TheCharlatan commented:
Updated 9de8b263dabd6dd2f86f1f0253c6ee3fac7a4407 -> dd290b3b58a1f9f9e182803866b8e2b9f72ecb3b ([preserveIndexOnRestart_1](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_1) -> [preserveIndexOnRestart_2](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/preserveIndexOnRestart_1..preserveIndexOnRestart_2))

* Added a commit for testing that the indexes are still there when continuing a reindex.

furszy commented on test/functional/feature_reindex.py:
```diff
@@ -73,13 +73,36 @@ def find_block(b, start):
+
+        # Restart node with reindex and stop reindex as soon as it starts reindexing
+        self.log.info("Restarting node while reindexing..")
+        with node.busy_wait_for_debug_log([b'initload thread start'], timeout=3):
```
thinking more about this, let's remove this timeout=3. It could fail on slow system or when running valgrind.

TheCharlatan commented on test/functional/feature_reindex.py:
```diff
@@ -73,13 +73,35 @@ def find_block(b, start):
+        node = self.nodes[0]
+        self.generate(node, 1500)
+        self.log.info("Wait until block filter index is synced..")
+        self.restart_node(0, ['-blockfilterindex'])
```
Yeah, I was thinking this might be used for some more regresssion tests, but as it is right now, it is useless. Will remove.

TheCharlatan commented:
Thanks for the reviews @maflcko,

891784ce73a1d911d33b125b66f3856fe6cda56b -> eeea0818c1a20adc5225b98b185953d386c033e0 ([preserveIndexOnRestart_3](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_3) -> [preserveIndexOnRestart_4](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/preserveIndexOnRestart_3..preserveIndexOnRestart_4))

* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1613029047), removed redundant initial syncing of the blockfilterindex.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1613028022), enforce using named argument.

cbergqvist commented on src/init.cpp:
```diff
@@ -1558,7 +1558,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         node::ChainstateLoadOptions options;
         options.mempool = Assert(node.mempool.get());
-        options.reindex = chainman.m_blockman.m_reindexing;
+        options.reindex = blockman_opts.reindex;
```
nit: How about taking the opportunity to nail down the mutability of `ChainstateLoadOptions` for improved readability?
```suggestion
        const node::ChainstateLoadOptions options{
            .mempool = Assert(node.mempool.get()),
            .reindex = blockman_opts.reindex,
            .reindex_chainstate = fReindexChainState,
            .prune = chainman.m_blockman.IsPruneMode(),
            .require_full_verification = args.IsArgSet("-checkblocks") || args.IsArgSet("-checklevel"),
            .check_blocks = args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS),
            .check_level = args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL),
            .coins_error_cb = [] {
                uiInterface.ThreadSafeMessageBox(
                    _("Error reading from database, shutting down."),
                    "", CClientUIInterface::MSG_ERROR);
            },
        };
```

cbergqvist reviewed (commented):
~~Tried cherry-picking only eeea0818c1a20adc5225b98b185953d386c033e0 on top of the base of the PR (058af75874ffa2b4064e3d6d30cc50f0ec754ba8) and ran the test and it succeeds without the other C++ changes?~~

Sorry for the false alarm, didn't rebuild before running the test. 🤦 Fails as expected now.

furszy reviewed (commented):
Code review ACK eeea0818c1

ryanofsky reviewed (approved):
Code review ACK eeea0818c1a20adc5225b98b185953d386c033e0. All the changes here look good. 

But I also think it was a mistake in #29817 to add the `BlockManagerOpts::reindex` option, and think it would be better to have a less confusing set of options. The following change built on top of this PR could provide a simpler alternative: 9c643e7ddd82523b84f65b614d3e6c1f640b23c7. Feel free to use any of these changes, or they could be a separate followup.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/30132#pullrequestreview-2094875458

> But I also think it was a mistake in https://github.com/bitcoin/bitcoin/pull/29817 to add the BlockManagerOpts::reindex option, and think it would be better to have a less confusing set of options. The following change built on top of this PR could provide a simpler alternative: https://github.com/bitcoin/bitcoin/commit/9c643e7ddd82523b84f65b614d3e6c1f640b23c7. Feel free to use any of these changes, or they could be a separate followup.

Removing the blockman option looks like a nice improvement, but I am not sure yet about the renaming. Going from `do_reindex` to `wipe_block_tree_db` back to `m_reindexing` seems a bit convoluted. I looked at adding a `do_reindex` argument to `BlockImport` instead so we don't have to set `m_reindexing` when loading the chainstate, but not sure if that is really better either.

In the commit message it also says:

> Also the previous set of options did not allow rebuilding the block database without also rebuilding the chainstate database, when it should be possible to do those independently.

This might be nice too. I was thinking a bit if this would really be safe, and I think it should be if the assumption anyway is that the chainstate is pristine.

stickies-v commented:
> The following change built on top of this PR could provide a simpler alternative: [9c643e7](https://github.com/bitcoin/bitcoin/commit/9c643e7ddd82523b84f65b614d3e6c1f640b23c7).

I like this suggestion. I think the naming is more clear in terms of the actual effect each variable has. I also think not having two `reindex` options is an improvement, the flow is more clear to me now.

I don't think `wipe_block_tree_db` (and a couple of other blockman options) should ultimately be a `ChainstateLoadOptions` member, but fixing that is probably beyond the scope of this PR (and I'm looking into doing that in a separate pull).

ryanofsky commented:
> Removing the blockman option looks like a nice improvement, but I am not sure yet about the renaming. Going from `do_reindex` to `wipe_block_tree_db` back to `m_reindexing` seems a bit convoluted.

That's a good point. One possible way to address it could be to rename `m_reindexing` as well like 38cc045b6966dda009332695ed55c86bd5a9fea3. If we move away from having multiple variables called "reindex" and instead name variables individually based on what they do, I think it would be an improvement overall.

The renames suggested in 9c643e7ddd82523b84f65b614d3e6c1f640b23c7 and 38cc045b6966dda009332695ed55c86bd5a9fea3 don't directly relate to this PR, so might make more sense as followups to avoid complicating things. On the other hand, if you took the portion of 9c643e7ddd82523b84f65b614d3e6c1f640b23c7 replacing `BlockManagerOpts::reindex` with `do_reindex` it would might make the third commit 9de8b263dabd6dd2f86f1f0253c6ee3fac7a4407 here easier to understand. Just feel free to ignore these suggestions for now or use any parts that seem useful.

> Also the previous set of options did not allow rebuilding the block database without also rebuilding the chainstate database, when it should be possible to do those independently.

Looking at the code more, I think setting `wipe_block_tree_db = true` and `wipe_chainstate_db = false` should probably work, even though we never do it. But if we add these options to the kernel API, the documentation could warn this combination is currently untested / unsupported. One problem with this combination is that it might not handle some reorgs, because `LoadExternalBlockFile` only currently scans block files and populates `CBlockIndex::nDataPos`. It doesn't currently scan undo files and populate `CBlockIndex::nUndoPos`, though it could.

theStack reviewed (approved):
ACK eeea0818c1a20adc5225b98b185953d386c033e0

Nice catch. Took me a while to see where b47bd959207e82555f07e028cc2246943d32d4c3 introduced the bug (the retry logic is confusing indeed!), but both the detailed commit messages and [yesterday's PR review club notes/log](https://bitcoincore.reviews/30132) were very helpful to grok it.

TheCharlatan commented:
Thanks for the review and ACKs, I will address the left over nits here shortly, I think re-ACKing should be easy enough.

ryanofsky committed:
kernel: Add less confusing reindex options

Drop confusing kernel options:

  BlockManagerOpts::reindex
  ChainstateLoadOptions::reindex
  ChainstateLoadOptions::reindex_chainstate

Replacing them with more straightforward options:

  ChainstateLoadOptions::wipe_block_tree_db
  ChainstateLoadOptions::wipe_chainstate_db

Having two options called "reindex" which did slightly different things
was needlessly confusing (one option wiped the block tree database, and
the other caused block files to be rescanned). Also the previous set of
options did not allow rebuilding the block database without also
rebuilding the chainstate database, when it should be possible to do
those independently.

ryanofsky committed:
blockman: Replace m_reindexing with m_blockfiles_indexed

This is a just a mechanical change, renaming and inverting the meaning
of the indexing variable.

"m_blockfiles_indexed" is a more straightforward name for this variable
because this variable just indicates whether or not
<datadir>/blocks/blk?????.dat files have been indexed in the
<datadir>/blocks/index LevelDB database. The name "m_reindexing" was
more confusing, it could be true even if -reindex was not specified, and
false when it was specified. Also, the previous name unnecessarily
required thinking about the whole reindexing process just to understand
simple checks in validation code about whether blocks were indexed.

The motivation for this change is to follow up on previous commits,
moving away from having multiple variables called "reindex" internally,
and instead naming variables individually after what they do and
represent.

TheCharlatan commented:
Thank you for your suggestions @ryanofsky! I applied both of your suggested patches.

eeea0818c1a20adc5225b98b185953d386c033e0 -> 682f1f1827dff78208bc1272f82d217c42a2cd69 ([preserveIndexOnRestart_4](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_4) -> [preserveIndexOnRestart_5](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/preserveIndexOnRestart_4..preserveIndexOnRestart_5))

* Addressed @theStack's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1630338668), using `chain_path` to construct the db path in the functional tests.
* Added @ryanofsky's [suggestion](https://github.com/bitcoin/bitcoin/pull/30132#pullrequestreview-2094875458) in this [patch](https://github.com/bitcoin/bitcoin/commit/9c643e7ddd82523b84f65b614d3e6c1f640b23c7), removing the blockmanager reindex option as well as renaming `reindex` to `wipe_block_tree_db` and `reindex_chainstate` to `wipe_chainstate_db` in the `ChainstateLoadOptions`.
* Added @ryanofsky's [suggestion](https://github.com/bitcoin/bitcoin/pull/30132#issuecomment-2147710137) in this [patch](https://github.com/bitcoin/bitcoin/commit/38cc045b6966dda009332695ed55c86bd5a9fea3), renaming `m_reindexing` to `m_blockfiles_indexed`.

stickies-v reviewed (commented):
Code LGTM 682f1f1827dff78208bc1272f82d217c42a2cd69 modulo [one small bug](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631416825) in comments. 

I couldn't find or think of any occasions where the new reduced wiping behaviour introduces problematic behaviour. The new variable names introduced make things significantly more clear and is a very welcome change.

stickies-v commented on src/init.cpp:
```diff
@@ -1694,7 +1696,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     // On first startup, warn on low block storage space
-    if (!chainman.m_blockman.m_reindexing && !fReindexChainState && chain_active_height <= 1) {
+    if (!chainman.m_blockman.m_blockfiles_indexed && !do_reindex_chainstate && chain_active_height <= 1) {
```
I actually don't understand why we also need to check for `!do_reindex_chainstate` here, when `CChain::vChain` is populated by the blockindex building process, so my intuition tells me `chain_active_height` should only be affected by `-reindex` (but lldb tells me otherwise, so it does indeed seem necessary).

ryanofsky replied:
> (but lldb tells me otherwise, so it does indeed seem necessary).

This happens because the initial chain tip is set based on the best block in the chainstate database, so if the database is being wiped, the tip is not set and the chain height is 0:

https://github.com/bitcoin/bitcoin/blob/6e4d18f37f42894fe9a0d7948c1da3f061fc6b60/src/node/chainstate.cpp#L136-L138

TheCharlatan commented:
Thanks for the review @stickies-v,

Updated f68cba29b3be0dec7877022b18a193a3b78c1099 -> f68cba29b3be0dec7877022b18a193a3b78c1099 ([preserveIndexOnRestart_5](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_5) -> [preserveIndexOnRestart_6](https://github.com/TheCharlatan/bitcoin/tree/preserveIndexOnRestart_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/preserveIndexOnRestart_5..preserveIndexOnRestart_6))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631387475), slightly changing the way the `wipe_chainstate_db` is set in `setup_common`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631413302), clarified `m_blockfiles_indexed` docstring.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631415537), changing `bitcoin-chainstate` log lines.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631416825), fixed missed renaming for the block storage check. The check should be done if a chainstate is present.

stickies-v reviewed (approved):
ACK f68cba29b3be0dec7877022b18a193a3b78c1099

fjahr commented on test/functional/feature_reindex.py:
```diff
@@ -73,13 +73,33 @@ def find_block(b, start):
+        self.log.info("Restarting node while reindexing..")
+        node.stop_node()
+        with node.busy_wait_for_debug_log([b'initload thread start']):
+            node.start(['-blockfilterindex', '-reindex'])
```
nit: Why was the blockfilterindex chosen here? I am assuming because it's slow? Would be good to add a comment because it may be confusing for others in the future what this choice has to do with the test.

TheCharlatan replied:
I just picked the one which furszy picked and I'm guessing he just picked one too. I'll check if there is any signifcant effect when picking a different one.

fjahr commented on test/functional/feature_reindex.py:
```diff
@@ -73,13 +73,33 @@ def find_block(b, start):
 
+    def continue_reindex_after_shutdown(self):
+        node = self.nodes[0]
+        self.generate(node, 1500)
```
nit: I guess this is needed so the node can be stopped fast enough. It's still a race and could turn out to be flakey in the CI, right? I don't have a better idea to fix this right now but a comment might be good to make this explicit and make future debugging easier if this turns out to be the case.

fjahr reviewed (commented):
Code review ACK f68cba29b3be0dec7877022b18a193a3b78c1099

I also confirmed that the new test fails before the changes here are applied.

I think the test could use a bit more explanation on its reasoning but this shouldn't block a merge of this as-is.

furszy reviewed (commented):
Code review ACK f68cba29b3be0dec7877022b18a193a3b78c1099

Great last commit.

Question about 804f09dfa116300914e2aeef05ed9710dd504e8c commit description:

> Also the previous set of
options did not allow rebuilding the block database without also
rebuilding the chainstate database, when it should be possible to do
those independently.

is this tested anywhere?

stickies-v commented on src/test/util/setup_common.cpp:
```diff
@@ -276,8 +276,8 @@ void ChainTestingSetup::LoadVerifyActivateChainstate()
     options.coins_db_in_memory = m_coins_db_in_memory;
-    options.reindex = chainman.m_blockman.m_reindexing;
-    options.reindex_chainstate = m_args.GetBoolArg("-reindex-chainstate", false);
+    options.wipe_block_tree_db = m_args.GetBoolArg("-reindex", false);
+    options.wipe_chainstate_db = m_args.GetBoolArg("-reindex", false) || m_args.GetBoolArg("-reindex-chainstate", false);
```
I suggested the current approach, see https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631387475

> I don't think this is used anywhere.

What do you mean?

furszy replied:
> I suggested the current approach, see https://github.com/bitcoin/bitcoin/pull/30132#discussion_r1631387475

Hmm ok. We should probably go further and deduplicate the init.cpp / setup_commons.cpp code somewhere in the future.

> > I don't think this is used anywhere.
> 
> What do you mean?

This is part of the unit test framework and no unit test, benchmark or fuzz test seems to make use of it.  "-reindex" and "-reindex-chainstate" are always unset.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/30132#pullrequestreview-2106239011

> is this tested anywhere?

No, and I'm not sure it should be given we don't support this. Maybe we can add a comment and assert that just wiping the block index db is not supported for now?

ryanofsky reviewed (approved):
Code review ACK f68cba29b3be0dec7877022b18a193a3b78c1099. Only changes since last review were cherry-picking suggested commits that rename variables, improving comments, and making some tweaks to test code.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30132#issuecomment-2157745629

> No, and I'm not sure it should be given we don't support this. Maybe we can add a comment and assert that just wiping the block index db is not supported for now?

FWIW, my original draft of 804f09dfa116300914e2aeef05ed9710dd504e8c added this code to LoadChainstate:

```c++
// For now, don't allow wiping block tree db without also wiping chainstate
// db. There's no reason this could not work in theory, but in practice the
// code path is untested, and to be really robust, the
// LoadExternalBlockFile function should to be updated to scan undo files,
// not just block files, and to populate CBlockIndex::nUndoPos, not just
// CBlockIndex::nDataPos.
assert(!options.wipe_block_tree_db || options.wipe_chainstate_db);
```

I decided to drop it to keep things simpler, since inevitably the kernel API will support combinations of options bitcoin core doesn't exercise or test, and it might be cumbersome to try to warn about all of them. But I could understand wanting to do it in some cases like this.

I think the PR is ready to merge, so you can let me know if you want to add an assert or just merge it in its current form.

TheCharlatan commented:
https://github.com/bitcoin/bitcoin/pull/30132#issuecomment-2158381286

> I think the PR is ready to merge, so you can let me know if you want to add an assert or just merge it in its current form.

I think this is rfm.
